File: src/activation.ts
import { Matrix } from "./Matrix";

export type ScalarActivationFN = {
  forward: (x: number) => number;
  backward: (x: number) => number;
};

const ScalarReLu: ScalarActivationFN = {
  forward: (x) => (x > 0 ? x : 0),
  backward: (x) => (x > 0 ? 1 : 0),
};

const ScalarTanH: ScalarActivationFN = {
  forward: TanhFn,
  backward: (x) => 1 - Math.pow(TanhFn(x), 2),
};

const ScalarSigmoid: ScalarActivationFN = {
  forward: SigmoidFn,
  backward: (x) => SigmoidFn(x) * (1 - SigmoidFn(x)),
};

export type ActivationFunctionKey = "sigmoid" | "tanh" | "relu" | "softmax";

export type ActivationFN = {
  forward: (sum: Matrix, isOutput: boolean) => Matrix;
  backward: (sum: Matrix) => Matrix;
  output: (output: Matrix, expected: Matrix) => Matrix;
};

export const ReLu: ActivationFN = {
  forward: (sum, isOutput) => activate(ScalarReLu.forward, sum, !isOutput),
  backward: (sum) => backpropagate(ScalarReLu.backward, sum),
  output: (sum) => backpropagate(ScalarReLu.backward, sum),
};

export const TanH: ActivationFN = {
  forward: (sum, isOutput) => activate(ScalarTanH.forward, sum, !isOutput),
  backward: (sum) => backpropagate(ScalarTanH.backward, sum),
  output: (sum) => backpropagate(ScalarTanH.backward, sum),
};

export const Sigmoid: ActivationFN = {
  forward: (sum, isOutput) => activate(ScalarSigmoid.forward, sum, !isOutput),
  backward: (sum) => backpropagate(ScalarSigmoid.backward, sum),
  output: (sum) => backpropagate(ScalarSigmoid.backward, sum),
};

export const Softmax: ActivationFN = {
  forward: (input, isOutput) => {
    const result = SoftmaxFn(input);
    if (!isOutput) {
      return result.unshift(1);
    }

    return result;
  },
  output: (sum, expected) => {
    return sum.subtract(expected);
  },
  backward: (sum) => {
    throw new Error("Use softmax only on output!");
    // const jacobianRows: number[][] = [];
    // const sums = SoftmaxFn(sum);
    // for (let i = 0; i < sum.M; i += 1) {
    //   jacobianRows[i] = [];
    //   const s_i = sums.get(i, 0);
    //   for (let j = 0; j < sum.M; j += 1) {
    //     if (i === j) {
    //       jacobianRows[i][j] = s_i * (1 - s_i);
    //     } else {
    //       jacobianRows[i][j] = -s_i * sums.get(j, 0);
    //     }
    //   }
    // }
    // const jacobian = Matrix.fromArrays(jacobianRows);
    // const result = multiplyer.transpose().multiply(jacobian).transpose();
    // return result;
  },
};

export const Identity: ActivationFN = {
  forward: (sum, isOutput) => activate((sum) => sum, sum, !isOutput),
  backward: (sum) => activate((sum) => sum, sum, false),
  output: (sum) => activate((sum) => sum, sum, false),
};

function backpropagate(fn: ScalarActivationFN["backward"], sum: Matrix) {
  return sum.map(fn);
}

function activate(
  fn: ScalarActivationFN["forward"],
  sum: Matrix,
  appendBias: boolean
) {
  const vect = sum.map(fn);
  if (appendBias) {
    return vect.unshift(1);
  }
  return vect;
}

function SigmoidFn(x: number) {
  const divisor = 1 + Math.exp(-x);
  return 1 / divisor;
}

function TanhFn(x: number) {
  return (Math.exp(x) - Math.exp(-x)) / (Math.exp(x) + Math.exp(-x));
}

function SoftmaxFn(vect: Matrix) {
  let max = vect.get(0, 0);
  vect.iterate((value) => {
    if (value > max) {
      max = value;
    }
  });

  let sum = 0;
  vect.iterate((value) => {
    sum += Math.exp(value - max);
  });

  return vect.map((x) => Math.exp(x - max) / sum);
}

File: rust/src/activation.rs
use crate::matrix::Matrix;

struct ScalarActivationFN {
    forward: fn(f64) -> f64,
    backward: fn(f64) -> f64,
}

pub enum ActivationFunctionKey {
    Sigmoid,
    TanH,
    ReLu,
    Softmax,
}

const SCALAR_RELU: ScalarActivationFN = ScalarActivationFN {
    forward: |x| -> f64 {
        if x > 0.0 {
            x
        } else {
            0.0
        }
    },
    backward: |x| -> f64 {
        if x > 0.0 {
            1.0
        } else {
            0.0
        }
    },
};

fn tanh_fn(x: f64) -> f64 {
    (x.exp() - (-x).exp()) / (x.exp() + (-x).exp())
}

const SCALAR_TANH: ScalarActivationFN = ScalarActivationFN {
    forward: tanh_fn,
    backward: |x| -> f64 { 1.0 - tanh_fn(x).powi(2) },
};

fn sigmoid_fn(x: f64) -> f64 {
    let divisor = 1.0 + (-x).exp();
    1.0 / divisor
}
const SCALAR_SIGMOID: ScalarActivationFN = ScalarActivationFN {
    forward: sigmoid_fn,
    backward: |x| -> f64 { sigmoid_fn(x) * (1.0 - sigmoid_fn(x)) },
};

pub struct ActivationFN {
    pub forward: fn(sum: &Matrix, isOutput: bool) -> Matrix,
    pub backward: fn(sum: &Matrix) -> Matrix,
    pub output: fn(output: &Matrix, expected: &Matrix) -> Matrix,
}

pub const RELU_ACTIVATION: ActivationFN = ActivationFN {
    forward: |sum, is_output| -> Matrix { activate(SCALAR_RELU.forward, sum, !is_output) },
    backward: |sum| -> Matrix { backpropagate(SCALAR_RELU.backward, sum) },
    output: |sum, _| -> Matrix { backpropagate(SCALAR_RELU.forward, sum) },
};

pub const TANH_ACTIVATION: ActivationFN = ActivationFN {
    forward: |sum, is_output| -> Matrix { activate(SCALAR_TANH.forward, sum, !is_output) },
    backward: |sum| -> Matrix { backpropagate(SCALAR_TANH.backward, sum) },
    output: |sum, _| -> Matrix { backpropagate(SCALAR_TANH.forward, sum) },
};

pub const SIGMOID_ACTIVATION: ActivationFN = ActivationFN {
    forward: |sum, is_output| -> Matrix { activate(SCALAR_SIGMOID.forward, sum, !is_output) },
    backward: |sum| -> Matrix { backpropagate(SCALAR_SIGMOID.backward, sum) },
    output: |sum, _| -> Matrix { backpropagate(SCALAR_SIGMOID.forward, sum) },
};

pub const IDENTITY_ACTIVATION: ActivationFN = ActivationFN {
    forward: |sum, is_output| -> Matrix { activate(|x| -> f64 { x }, sum, !is_output) },
    backward: |_| -> Matrix { panic!("use identity only on first layer") },
    output: |_, _| -> Matrix { panic!("use identity only on first layer") },
};

pub const SOFTMAX_ACTIVATION: ActivationFN = ActivationFN {
    forward: |sum, is_output| -> Matrix {
        let result = softmax_fn(sum);
        if !is_output {
            return result.unshift(1.0);
        }
        result
    },
    output: |sum, expected| -> Matrix { sum.subtract(expected) },
    backward: |_| -> Matrix { panic!("use softmax only on output") },
};

fn activate(mapper: fn(f64) -> f64, sum: &Matrix, append_bias: bool) -> Matrix {
    let vect = sum.map(&|x, _, _| -> f64 { mapper(x) });
    if append_bias {
        return vect.unshift(1.0);
    }
    vect
}

fn backpropagate(mapper: fn(f64) -> f64, sum: &Matrix) -> Matrix {
    sum.map(&|x, _, _| -> f64 { mapper(x) })
}

fn softmax_fn(vector: &Matrix) -> Matrix {
    let mut max = vector.get(0, 0);
    vector.iterate(&mut |value, _| {
        if value > max {
            max = value
        }
    });

    let mut sum = 0.0;
    vector.iterate(&mut |value, _| {
        sum += (value - max).exp();
    });

    vector.map(&|x, _, _| -> f64 { (x - max).exp() / sum })
}
